vowpalwabbit/topk.cc

```diff
@@ -5,51 +5,77 @@ license as described in the file LICENSE.
  */
 #include <float.h>
 #include <sstream>
-#include <queue>
 
-#include "reductions.h"
+#include "topk.h"
 #include "vw.h"
 
-using namespace std;
 using namespace VW::config;
 
-using scored_example = pair<float, v_array<char>>;
+VW::topk::topk(uint32_t K) : m_K(K) {}
 
-struct compare_scored_examples
+void VW::topk::predict(LEARNER::single_learner& base, multi_ex& ec_seq)
 {
-  constexpr bool operator()(scored_example const& a, scored_example const& b) const { return a.first > b.first; }
-};
+  for (auto ec : ec_seq)
+  {
+    base.predict(*ec);
+    update_priority_queue(ec->pred.scalar, ec->tag);
+  }
+}
 
-struct topk
+void VW::topk::learn(LEARNER::single_learner& base, multi_ex& ec_seq)
 {
-  uint32_t K;  // rec number
-  priority_queue<scored_example, vector<scored_example>, compare_scored_examples> pr_queue;
-  vw* all;
-};
+  for (auto ec : ec_seq)
+  {
+    base.learn(*ec);
+    update_priority_queue(ec->pred.scalar, ec->tag);
+  }
+}
 
-void print_result(int f, priority_queue<scored_example, vector<scored_example>, compare_scored_examples>& pr_queue)
+void VW::topk::update_priority_queue(float pred, v_array<char>& tag)
 {
-  if (f >= 0)
+  if (m_pr_queue.size() < m_K)
+    m_pr_queue.push(std::make_pair(pred, tag));
+  else if (m_pr_queue.top().first < pred)
+  {
+    m_pr_queue.pop();
+    m_pr_queue.push(std::make_pair(pred, tag));
+  }
+}
+
+std::vector<VW::topk::scored_example> VW::topk::drain_queue()
+{
+  std::vector<scored_example> result;
```

Comment:
Won't this cause allocations for every finish_example?

Reply:
This was the original reason I suggested the multimap. priority queue is fine for the top element, like you mentioned above, but this was bad

Reply:
Yep when using a pr it has to copy out into a vector